core/src/main/java/com/datastax/oss/driver/api/core/AllNodesFailedException.java

```diff
@@ -17,87 +17,150 @@
+import java.util.Map.Entry;
+import java.util.Set;
 
 /**
  * Thrown when a query failed on all the coordinators it was tried on. This exception may wrap
  * multiple errors, use {@link #getErrors()} to inspect the individual problem on each node.
  */
 public class AllNodesFailedException extends DriverException {
 
+  /** @deprecated Use {@link #fromErrors(List)} instead. */
   @NonNull
+  @Deprecated
   public static AllNodesFailedException fromErrors(@Nullable Map<Node, Throwable> errors) {
     if (errors == null || errors.isEmpty()) {
       return new NoNodeAvailableException();
     } else {
-      return new AllNodesFailedException(ImmutableMap.copyOf(errors));
+      return new AllNodesFailedException(groupByNode(errors.entrySet()));
     }
   }
 
   @NonNull
-  public static AllNodesFailedException fromErrors(
-      @Nullable List<Map.Entry<Node, Throwable>> errors) {
-    Map<Node, Throwable> map;
+  public static AllNodesFailedException fromErrors(@Nullable List<Entry<Node, Throwable>> errors) {
     if (errors == null || errors.isEmpty()) {
-      map = null;
+      return new NoNodeAvailableException();
     } else {
-      ImmutableMap.Builder<Node, Throwable> builder = ImmutableMap.builder();
-      for (Map.Entry<Node, Throwable> entry : errors) {
-        builder.put(entry);
-      }
-      map = builder.build();
+      return new AllNodesFailedException(groupByNode(errors));
     }
-    return fromErrors(map);
   }
 
-  private final Map<Node, Throwable> errors;
+  private final Map<Node, List<Throwable>> errors;
 
+  /** @deprecated Use {@link #AllNodesFailedException(String, ExecutionInfo, Iterable)} instead. */
+  @Deprecated
   protected AllNodesFailedException(
       @NonNull String message,
       @Nullable ExecutionInfo executionInfo,
       @NonNull Map<Node, Throwable> errors) {
+    this(message, executionInfo, groupByNode(errors.entrySet()));
+  }
+
+  protected AllNodesFailedException(
+      @NonNull String message,
+      @Nullable ExecutionInfo executionInfo,
+      @NonNull Iterable<Entry<Node, List<Throwable>>> errors) {
     super(message, executionInfo, null, true);
-    this.errors = errors;
+    this.errors = toDeepImmutableMap(errors);
   }
 
-  private AllNodesFailedException(Map<Node, Throwable> errors) {
+  private AllNodesFailedException(Collection<Entry<Node, List<Throwable>>> errors) {
     this(
         buildMessage(
             String.format("All %d node(s) tried for the query failed", errors.size()), errors),
         null,
         errors);
   }
 
-  private static String buildMessage(String baseMessage, Map<Node, Throwable> errors) {
+  private static String buildMessage(
+      String baseMessage, Collection<Entry<Node, List<Throwable>>> errors) {
     int limit = Math.min(errors.size(), 3);
-    String details =
-        Joiner.on(", ").withKeyValueSeparator(": ").join(Iterables.limit(errors.entrySet(), limit));
-
+    Iterator<Entry<Node, List<Throwable>>> iterator = errors.iterator();
```

Comment:
We've lost the limit in the refactoring. You can use `Iterables.limit()` on this line.

Reply:
The only thing that bothers me is that the signature of this method, at least in theory, cannot guarantee that exactly N distinct nodes will be printed, since duplicate entries could happen (not in the current implementation, but if someone extends this class, then it's not impossible). I will try to change the signature to accept a map of nodes instead, just to be on the safe side.